client/client-multi/client-c/src/kaa/platform-impl/posix/key_utils.c

```diff
@@ -0,0 +1,417 @@
+    mbedtls_entropy_free(&entropy);
+
+    return ret;
+}
+
+static kaa_error_t set_rsa_public_key(mbedtls_pk_context *pk, const uint8_t *key, size_t key_size)
+{
+    if (!pk || !key || !key_size) {
+        return KAA_ERR_BADPARAM;
+    }
+
+    if (mbedtls_pk_parse_public_key(pk, key, key_size)!= 0) {
+        return KAA_ERR_INVALID_PUB_KEY;
+    }
+
+    return KAA_ERR_NONE;
+}
+
+static kaa_error_t aes_encrypt_decrypt(int mode, const uint8_t *input, size_t input_size,
+        uint8_t *output, const uint8_t *key)
+{
+    if ((input_size % 16) != 0 || !input) {
+        return KAA_ERR_BADPARAM;
+    }
+
+    if (mode != MBEDTLS_AES_ENCRYPT && mode != MBEDTLS_AES_DECRYPT) {
+        return KAA_ERR_BADPARAM;
+    }
+
+    static bool initialized = false;
+    static mbedtls_aes_context aes_ctx;
+
+    if (!initialized) {
+        mbedtls_aes_init(&aes_ctx);
+        initialized = true;
+    }
+
+    if (mode == MBEDTLS_AES_ENCRYPT) {
+        mbedtls_aes_setkey_enc(&aes_ctx, key, KAA_SESSION_KEY_LENGTH * 8);
+        mbedtls_aes_crypt_ecb(&aes_ctx, MBEDTLS_AES_ENCRYPT, input, output);
+    } else if (mode == MBEDTLS_AES_DECRYPT) {
+        mbedtls_aes_setkey_dec(&aes_ctx, key, KAA_SESSION_KEY_LENGTH * 8);
+        mbedtls_aes_crypt_ecb(&aes_ctx, MBEDTLS_AES_DECRYPT, input, output);
+    }
+
+    return KAA_ERR_NONE;
+}
+
+static int init_aes_key(unsigned char *key, size_t bytes)
+{
+    mbedtls_ctr_drbg_context ctr_drbg;
+    mbedtls_entropy_context entropy;
+
+    const uint8_t pers[] = "aes_generate_key";
+    int ret;
+
+    mbedtls_entropy_init(&entropy);
+    mbedtls_ctr_drbg_init(&ctr_drbg);
+
+    if ((ret = mbedtls_ctr_drbg_seed(&ctr_drbg, mbedtls_entropy_func, &entropy,
+            pers, sizeof(pers) - 1)) == 0) {
+        ret = mbedtls_ctr_drbg_random(&ctr_drbg, key, bytes);
+    }
+
+    mbedtls_ctr_drbg_free(&ctr_drbg);
+    mbedtls_entropy_free(&entropy);
+
+    return ret;
+}
+
+/* Performs initialization of the keys */
+kaa_error_t kaa_init_keys(void)
+{
+    /* Initialization should be performed only once */
+    static bool initialized = false;
+    if (!initialized) {
+        /* Generate session key */
+        if (init_aes_key(keys.session_key, KAA_SESSION_KEY_LENGTH)) {
+            return KAA_ERR_BADDATA;
+        }
+        keys.session_key_length = KAA_SESSION_KEY_LENGTH;
+        if (mbedtls_pk_parse_keyfile(&pk_context_, KAA_PRIVATE_KEY_STORAGE, NULL)) {
+#ifdef KAA_RUNTIME_KEY_GENERATION
+            if (rsa_genkey(&pk_context_)) {
+                return -1;
+            }
+#else
+            return KAA_ERR_BADDATA;
+#endif
+        }
+         if (mbedtls_pk_parse_public_keyfile(&pk_pub_context, KAA_PUBLIC_KEY_STORAGE)) {
+#ifdef KAA_RUNTIME_KEY_GENERATION
+		pk_pub_context = pk_context_;
+#else
+            return KAA_ERR_BADDATA;
+#endif
+        }
+        initialized = true;
+    }
+
+    return KAA_ERR_NONE;
+}
+
+void  kaa_deinit_keys(void)
+{
+    mbedtls_pk_free(&pk_context_);
```

Comment:
What's about `pk_pub_context`? Investigate, if `mbedtls_pk_free` should be called against public key context.